File: JavaSwarms/SwarmChemistry_sourcecode/src/swarm/engine/Recipe.java
package swarm.engine;

// Recipe.java
//
// Part of:
// SwarmChemistry.java
// Public Release Version 1.2.0
//
// 2006-2009 (c) Copyright by Hiroki Sayama. All rights reserved.
//
// Send any correspondences to:
//   Hiroki Sayama, D.Sc.
//   Assistant Professor, Department of Bioengineering
//   Binghamton University, State University of New York
//   P.O. Box 6000, Binghamton, NY 13902-6000, USA
//   Tel: +1-607-777-4439  Fax: +1-607-777-5780
//   Email: sayama@binghamton.edu
//
// For more information about this software, see:
//   http://bingweb.binghamton.edu/~sayama/SwarmChemistry/
//

import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;


public class Recipe {

	private ArrayList<Parameters> parameters;
	private ArrayList<Integer> popCounts;
	private String recipeText;

	private static final double populationChangeMagnitude = 0.8;
	private static final double duplicationOrDeletionRatePerParameterSets = 0.1;
	private static final double randomAdditionRatePerRecipe = 0.1;
	private static final double pointMutationRatePerParameter = 0.1;
	private static final double pointMutationMagnitude = 0.5;

	public Recipe(String text) {
		setFromText(text);
	}

	public Recipe(List<Individual> sol) {
		setFromPopulation(sol);
	}

	public boolean setFromText(String text) {
		char ch;
		int numberOfIngredients, numberOfIndividuals;
		double neighborhoodRadius, normalSpeed, maxSpeed, c1, c2, c3, c4, c5;

		StringBuffer recipeProcessed = new StringBuffer(text.length());
		for (int i = 0; i < text.length(); i++) {
			ch = text.charAt(i);
			if ((ch >= '0' && ch <= '9') || (ch == '.'))
				recipeProcessed.append(ch);
			else if (recipeProcessed.length() > 0) {
				if (recipeProcessed.charAt(recipeProcessed.length() - 1) != ' ')
					recipeProcessed.append(' ');
			}
		}

		StringTokenizer st = new StringTokenizer(recipeProcessed.toString(),
				" ");

		if (st.countTokens() % 9 != 0) {
			recipeText = "*** Formatting error ***\n" + text;
			parameters = null;
			popCounts = null;
			return false;
		}

		numberOfIngredients = st.countTokens() / 9;
		if (numberOfIngredients == 0) {
			recipeText = "*** No ingredients ***\n" + text;
			parameters = null;
			popCounts = null;
			return false;
		}
		if (numberOfIngredients > Parameters.numberOfIndividualsMax)
			numberOfIngredients = Parameters.numberOfIndividualsMax;

		parameters = new ArrayList<Parameters>();
		popCounts = new ArrayList<Integer>();

		try {
			for (int i = 0; i < numberOfIngredients; i++) {
				numberOfIndividuals = Integer.parseInt(st.nextToken());
				if (numberOfIndividuals < 1)
					numberOfIndividuals = 1;
				st.nextToken();
				st.nextToken();
				st.nextToken();
				st.nextToken();
				st.nextToken();
				st.nextToken();
				st.nextToken();
				st.nextToken();
			}

			st = new StringTokenizer(recipeProcessed.toString(), " ");

			for (int i = 0; i < numberOfIngredients; i++) {
				numberOfIndividuals = Integer.parseInt(st.nextToken());
				if (numberOfIndividuals < 1)
					numberOfIndividuals = 1;
				neighborhoodRadius = Double.parseDouble(st.nextToken());
				normalSpeed = Double.parseDouble(st.nextToken());
				maxSpeed = Double.parseDouble(st.nextToken());
				c1 = Double.parseDouble(st.nextToken());
				c2 = Double.parseDouble(st.nextToken());
				c3 = Double.parseDouble(st.nextToken());
				c4 = Double.parseDouble(st.nextToken());
				c5 = Double.parseDouble(st.nextToken());
				parameters.add(new Parameters(neighborhoodRadius,
						normalSpeed, maxSpeed, c1, c2, c3, c4, c5));
				popCounts.add(new Integer(numberOfIndividuals));
			}
		} catch (NumberFormatException nfe) {
			recipeText = "*** Formatting error ***\n" + text;
			parameters = null;
			popCounts = null;
			return false;
		}

		boundPopulationSize();
		setFromPopulation(createPopulation(300, 300));
		return true;
	}

	public void boundPopulationSize() {
		int totalPopulation = 0;
		double rescalingRatio;

		int numberOfIngredients = parameters.size();

		for (int i = 0; i < numberOfIngredients; i++)
			totalPopulation += popCounts.get(i).intValue();

		if (totalPopulation > Parameters.numberOfIndividualsMax)
			rescalingRatio = (double) (Parameters.numberOfIndividualsMax - numberOfIngredients)
					/ (totalPopulation == numberOfIngredients ? 1.0
							: (double) (totalPopulation - numberOfIngredients));

		else
			rescalingRatio = 1;

		if (rescalingRatio != 1)
			for (int i = 0; i < numberOfIngredients; i++)
				popCounts.set(i, new Integer(1 + (int) Math
						.floor((double) (popCounts.get(i).intValue() - 1)
								* rescalingRatio)));
	}

	public void setFromPopulation(List<Individual> sol) {
		parameters = new ArrayList<Parameters>();
		popCounts = new ArrayList<Integer>();

		Parameters tempParam;

		for (int i = 0; i < sol.size(); i++) {
			tempParam = sol.get(i).getGenome();

			boolean alreadyInParameters = false;
			for (int j = 0; j < parameters.size(); j++) {
				if (parameters.get(j).equals(tempParam)) {
					alreadyInParameters = true;
					popCounts.set(j, new Integer(
							popCounts.get(j).intValue() + 1));
				}
			}
			if (alreadyInParameters == false) {
				parameters.add(tempParam);
				popCounts.add(new Integer(1));
			}
		}

		setRecipeText();
	}

	private void setRecipeText() {
		Parameters tempParam;

		recipeText = "";

		for (int i = 0; i < parameters.size(); i++) {
			tempParam = parameters.get(i);
			recipeText += "" + popCounts.get(i).intValue() + " * ("
					+ shorten(tempParam.getNeighborhoodRadius()) + ", "
					+ shorten(tempParam.getNormalSpeed()) + ", "
					+ shorten(tempParam.getMaxSpeed()) + ", "
					+ shorten(tempParam.getC1()) + ", " + shorten(tempParam.getC2())
					+ ", " + shorten(tempParam.getC3()) + ", "
					+ shorten(tempParam.getC4()) + ", " + shorten(tempParam.getC5())
					+ ")\n";
		}
	}

	private double shorten(double d) {
		return Math.round(d * 100.0) / 100.0;
	}

	public ArrayList<Individual> createPopulation(int width, int height) {
		if (parameters == null)
			return null;

		ArrayList<Individual> newPopulation = new ArrayList<Individual>();
		Parameters tempParam;

		for (int i = 0; i < parameters.size(); i++) {
			tempParam = parameters.get(i);
			for (int j = 0; j < popCounts.get(i).intValue(); j++)
				newPopulation.add(new Individual(Math.random() * width,
						Math.random() * height, Math.random() * 10 - 5, Math
								.random() * 10 - 5, new Parameters(
								tempParam)));
		}

		return newPopulation;
	}
	 
	public boolean mutate() {
		boolean mutated = false;
		int numberOfIngredients = parameters.size();

		// Insertions, duplications and deletions

		for (int j = 0; j < numberOfIngredients; j++) {
			if (Math.random() < duplicationOrDeletionRatePerParameterSets) {
				if (Math.random() < .5) { // Duplication
					mutated = true;
					parameters.add(j + 1, parameters
							.get(j));
					popCounts.add(j + 1, popCounts
							.get(j));
					numberOfIngredients++;
					j++;
				} else { // Deletion
					if (numberOfIngredients > 1) {
						mutated = true;
						parameters.remove(j);
						popCounts.remove(j);
						numberOfIngredients--;
						j--;
					}
				}
			}
		}

		if (Math.random() < randomAdditionRatePerRecipe) { // Addition
			mutated = true;
			parameters.add(new Parameters());
			popCounts.add(new Integer((int) (Math.random()
					* Parameters.numberOfIndividualsMax * 0.5) + 1));
		}

		// Then Point Mutations

		Parameters tempParam;

		for (int j = 0; j < numberOfIngredients; j++) {
			tempParam = new Parameters(parameters.get(j));
			tempParam.inducePointMutations(pointMutationRatePerParameter,
					pointMutationMagnitude);
			if (!parameters.get(j).equals(tempParam)) {
				mutated = true;
				parameters.set(j, tempParam);
			}
		}

		boundPopulationSize();
		
		return mutated;
	}
	
	public String getRecipeText() {
		setRecipeText();
		return recipeText;
	}
}

File: XNASwarms/SwarmEngine/Recipes/Recipe.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SwarmEngine
{
    public class Recipe
    {

        private List<Parameters> parameters;
        private List<int> popCounts;

        public string Message
        {
            get;
            private set;
        }

        private static double populationChangeMagnitude = 0.8;
        private static double duplicationOrDeletionRatePerParameterSets = 0.1;
        private static double randomAdditionRatePerRecipe = 0.1;
        private static double pointMutationRatePerParameter = 0.1;
        private static double pointMutationMagnitude = 0.5;

        Random rand = new Random();

        public Recipe(String text)
        {
            SetFromText(text);
        }

        public Recipe(List<Species> sol)
        {
            SetFromPopulation(sol);
        }

        public bool SetFromText(String text)
        {
            char ch;
            int numberOfIngredients, numberOfIndividuals;
            double neighborhoodRadius, normalSpeed, maxSpeed, c1, c2, c3, c4, c5;

            StringBuilder recipeProcessed = new StringBuilder(text.Length);
            for (int i = 0; i < text.Length; i++)
            {
                ch = text[i];
                if ((ch >= '0' && ch <= '9') || (ch == '.'))
                    recipeProcessed.Append(ch);
                else if (recipeProcessed.Length > 0)
                {
                    if (recipeProcessed[recipeProcessed.Length - 1] != ' ')
                        recipeProcessed.Append(' ');
                }
            }

            string[] st = recipeProcessed.ToString().Split(' ');


            if (st.Length % 9 != 0)
            {
                Message = "*** Formatting error ***\n" + text;
                parameters = null;
                popCounts = null;
                return false;
            }

            numberOfIngredients = st.Length / 9;
            if (numberOfIngredients == 0)
            {
                Message = "*** No ingredients ***\n" + text;
                parameters = null;
                popCounts = null;
                return false;
            }
            if (numberOfIngredients > StaticWorldParameters.numberOfIndividualsMax)
                numberOfIngredients = StaticWorldParameters.numberOfIndividualsMax;

            parameters = new List<Parameters>();
            popCounts = new List<int>();

            try
            {
                for (int i = 0; i < numberOfIngredients; i++)
                {
                    int x = int.Parse(st[0].ToString().Replace("\"", ""));
                    numberOfIndividuals = int.Parse(st[0].ToString());
                    if (numberOfIndividuals < 1)
                        numberOfIndividuals = 1;
                    neighborhoodRadius = Double.Parse(st[1]);
                    normalSpeed = Double.Parse(st[2]);
                    maxSpeed = Double.Parse(st[3]);
                    c1 = Double.Parse(st[4]);
                    c2 = Double.Parse(st[5]);
                    c3 = Double.Parse(st[6]);
                    c4 = Double.Parse(st[7]);
                    c5 = Double.Parse(st[8]);
                    parameters.Add(new Parameters(neighborhoodRadius,
                            normalSpeed, maxSpeed, c1, c2, c3, c4, c5));
                    popCounts.Add(numberOfIndividuals);
                }
            }
            catch (Exception nfe)
            {
                //TODO : Find a better Exception for this catch
                Message = "*** Formatting error ***\n" + text;
                parameters = null;
                popCounts = null;
                return false;
            }

            BoundPopulationSize();
            return true;
        }

        public void BoundPopulationSize()
        {
            double totalPopulation = 0;
            double rescalingRatio;

            int numberOfIngredients = parameters.Count;

            for (int i = 0; i < numberOfIngredients; i++)
                totalPopulation += popCounts[i];

            if (totalPopulation > StaticWorldParameters.numberOfIndividualsMax)
                rescalingRatio = (double)(StaticWorldParameters.numberOfIndividualsMax - numberOfIngredients)
                        / (totalPopulation == numberOfIngredients ? 1.0
                                : (double)(totalPopulation - numberOfIngredients));

            else
                rescalingRatio = 1;

            if (rescalingRatio != 1)
                for (int i = 0; i < numberOfIngredients; i++)
                    popCounts[i] = (1 + (int)Math.Floor((double)(popCounts[i] - 1) * rescalingRatio));
        }

        public void SetFromPopulation(List<Species> species)
        {
            parameters = new List<Parameters>();
            popCounts = new List<int>();

            Parameters tempParam;

            for (int i = 0; i < species.Count(); i++)
            {

                for (int j = 0; j < species[i].Count(); j++)
                {
                    tempParam = species[i][j].Genome;

                    bool alreadyInParameters = false;
                    for (int k = 0; k < parameters.Count; k++)
                    {
                        if (parameters[k].equals(tempParam))
                        {
                            alreadyInParameters = true;
                            popCounts[k] = (popCounts[k] + 1);
                        }
                    }
                    if (alreadyInParameters == false)
                    {
                        parameters.Add(tempParam);
                        popCounts.Add(1);
                    }
                }
            }

            SetRecipeText();
        }

        private void SetRecipeText()
        {
            Parameters tempParam;

            Message = "";

            for (int i = 0; i < parameters.Count(); i++)
            {
                tempParam = parameters[i];
                Message += "" + popCounts[i] + " * ("
                        + Shorten(tempParam.getNeighborhoodRadius()) + ", "
                        + Shorten(tempParam.getNormalSpeed()) + ", "
                        + Shorten(tempParam.getMaxSpeed()) + ", "
                        + Shorten(tempParam.getC1()) + ", " + Shorten(tempParam.getC2())
                        + ", " + Shorten(tempParam.getC3()) + ", "
                        + Shorten(tempParam.getC4()) + ", " + Shorten(tempParam.getC5())
                        + ")\n";
            }
        }

        private double Shorten(double d)
        {
            return Math.Round(d * 100.0) / 100.0;
        }

        public List<Species> CreatePopulation(int width, int height)
        {
            if (parameters == null)
                return null;

            List<Species> newListSpecies = new List<Species>();
            Parameters tempParam;

            for (int i = 0; i < parameters.Count; i++)
            {
                tempParam = parameters[i];
                var spcs = new Species(new List<Individual>());
                for (int j = 1; j < popCounts[i]; j++)
                {
                    spcs.Add(new Individual(j, rand.NextDouble() * width,
                            rand.NextDouble() * height, rand.NextDouble() * 10 - 5,
                            rand.NextDouble() * 10 - 5, new Parameters(
                                    tempParam)));

                }
                newListSpecies.Add(spcs);
            }
            //mutate();
            return newListSpecies;
        }

        //public bool mutate() {
        //    bool mutated = false;
        //    int numberOfIngredients = parameters.Count();

        //    // Insertions, duplications and deletions

        //    for (int j = 0; j < numberOfIngredients-1; j++) {
        //        if (rand.NextDouble() < duplicationOrDeletionRatePerParameterSets)
        //        {
        //            if (rand.NextDouble() < .5) { // Duplication
        //                mutated = true;
        //                parameters[j + 1] = parameters[j];
        //                popCounts[j + 1] = popCounts[j];
        //                numberOfIngredients++;
        //                j++;
        //            } else { // Deletion
        //                if (numberOfIngredients > 1) {
        //                    mutated = true;
        //                    parameters.RemoveAt(j);
        //                    popCounts.RemoveAt(j);
        //                    numberOfIngredients--;
        //                    j--;
        //                }
        //            }
        //        }
        //    }

        //    if (rand.NextDouble() < randomAdditionRatePerRecipe)
        //    { // Addition
        //        mutated = true;
        //        parameters.Add(new Parameters());
        //        popCounts.Add((int)(rand.NextDouble() * Parameters.numberOfIndividualsMax * 0.5) + 1);
        //    }

        //    // Then Point Mutations

        //    Parameters tempParam;

        //    for (int j = 0; j < numberOfIngredients; j++) {
        //        tempParam = new Parameters(parameters[j]);
        //        tempParam.inducePointMutations(pointMutationRatePerParameter,
        //                pointMutationMagnitude);
        //        if (!parameters[j].equals(tempParam)) {
        //            mutated = true;
        //            parameters[j] = tempParam;
        //        }
        //    }

        //    boundPopulationSize();

        //    return mutated;
        //}

    }
}
